### PR TITLE
Bug 1813252 - Refactor MessagingMiddlewareTest to use a real Store in…

### PR DESCRIPTION
**Pull request by sarah541**
…stance

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1813252

### CODE REVIEW

File: fenix/app/src/test/java/org/mozilla/fenix/gleanplumb/state/MessagingMiddlewareTest.kt [hunk 1]
```diff
@@ -178,27 +195,36 @@ class MessagingMiddlewareTest {
...
-    fun `WHEN onMessageDismissed THEN updateMetadata,removeMessage , UpdateMessages and removeMessageToShowIfNeeded`() = runTestOnMain {
+    fun `WHEN onMessageDismissed THEN updateMetadata,removeMessage , UpdateMessages and consumeMessageToShowIfNeeded`() = runTestOnMain {
         val message = Message(
             "control-id",
             mockk(relaxed = true),
             action = "action",
             mockk(relaxed = true),
             listOf("trigger"),
-            Message.Metadata("same-id"),
+            Message.Metadata("same-id", displayCount = 0),
```
Comment (jonalmeida) [inline fenix/app/src/test/java/org/mozilla/fenix/gleanplumb/state/MessagingMiddlewareTest.kt:205]: For `onMessageDismissed` to happen, wouldn't we need to have displayed the message at least once?

```suggestion
            Message.Metadata("same-id", displayCount = 1),
```

File: fenix/app/src/test/java/org/mozilla/fenix/gleanplumb/state/MessagingMiddlewareTest.kt [hunk 2]
```diff
@@ -178,27 +195,36 @@ class MessagingMiddlewareTest {
...
+        store.waitUntilIdle()
 
-        coVerify { messagingController.onMessageDismissed(message.metadata) }
-        coVerify { messagingStorage.updateMetadata(message.metadata.copy(dismissed = true)) }
-        verify { middlewareContext.dispatch(UpdateMessages(emptyList())) }
-        verify { spiedMiddleware.removeMessage(middlewareContext, message) }
+        // removeMessages causes messages size to be 0
+        assertEquals(0, store.state.messaging.messages.size)
+        // consumeMessageToShowIfNeeded causes messageToShow map size to be 0
+        assertEquals(0, store.state.messaging.messageToShow.size)
```
Comment (jonalmeida) [inline fenix/app/src/test/java/org/mozilla/fenix/gleanplumb/state/MessagingMiddlewareTest.kt:227]: The assertion here is right, but if you look up at your state, you start off with `messageToShow` as empty. So we're asserting here that nothing has changed? Shouldn't we have had a `messageToShow` first, then after the dismissal, the message is removed from `messages` and `messageToShow`?

Comment (jonalmeida) [reply fenix/app/src/test/java/org/mozilla/fenix/gleanplumb/state/MessagingMiddlewareTest.kt:227]: Let's add an assertion before the `dispatch(MessageDismissed(message))` that the `messageToShow` has a message in there first.

Comment (sarah541) [reply fenix/app/src/test/java/org/mozilla/fenix/gleanplumb/state/MessagingMiddlewareTest.kt:199]: Yes, totally agree here

File: fenix/app/src/test/java/org/mozilla/fenix/gleanplumb/state/MessagingMiddlewareTest.kt [hunk 3]
```diff
@@ -48,240 +38,226 @@ class MessagingMiddlewareTest {
...
 
-        coEvery { messagingStorage.getMessages() } returns messages
+        val message = Message(
+            "message1",
+            mockk(relaxed = true),
+            "",
+            StyleData(),
+            emptyList(),
+            Message.Metadata("same-id", displayCount = 0),
+        )
```
Comment (jonalmeida) [inline fenix/app/src/test/java/org/mozilla/fenix/gleanplumb/state/MessagingMiddlewareTest.kt:73]: So here is the other bit of refactoring we could do if you think it's worthwhile.

We create virtually identical messages for each test with only one/two changes. A nice way to make these tests easier to parse or easier to write more in the future is with simple helper functions for our tests:

```kotlin
private fun createMessage(
    messageId: String = "control-id",
    data: MessageData = mockk(relaxed = true),
    action: String = "action",
    styleData: StyleData = StyleData(),
    triggers: List<String> = listOf("triggers"),
    metadata: Message.Metadata = createMetadata(),
) = Message(messageId, data, action, styleData, triggers, metadata)

private fun createMetadata(
    id: String = "same-id",
    displayCount: Int = 0,
    pressed: Boolean = false,
    dismissed: Boolean = false,
    lastTimeShown: Long = 0L,
    latestBootIdentifier: String? = null,
) = Message.Metadata(
    id,
    displayCount,
    pressed,
    dismissed,
    lastTimeShown,
    latestBootIdentifier,
)
```

This means that this line or the ones below turn into one/two liners like this:

```suggestion
        val message = createMessage()
```

**Review by jonalmeida — APPROVED** [review]
Great! This looks good to me to land.

I left some nits which aren't blocking but worth considering.

cc: @t-p-white who has worked on messaging and might want to know that we've updated these tests to properly utilizee the messaging middleware code in them.

File: fenix/app/src/test/java/org/mozilla/fenix/gleanplumb/state/MessagingMiddlewareTest.kt (hunk 3 above)
Comment (sarah541) [reply fenix/app/src/test/java/org/mozilla/fenix/gleanplumb/state/MessagingMiddlewareTest.kt]: Yes makes sense

**Pull request merged**